var/spack/repos/builtin/packages/babelstream/package.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+from spack.package import *
+
+
+class Babelstream(CMakePackage):
+    """Measure memory transfer rates to/from global device memory on GPUs.
+    This benchmark is similar in spirit, and based on, the STREAM benchmark for CPUs."""
+
+    homepage = "https://github.com/UoB-HPC/BabelStream"
+    git = "https://github.com/UoB-HPC/BabelStream.git"
+    url = "https://github.com/UoB-HPC/BabelStream/archive/refs/tags/v4.0.tar.gz"
+
+    maintainers = ["rob64j", "tomdeakin", "tom91136"]
+
+    version("main", branch="main")
+    version("develop", branch="develop")
+    version("4.0", sha256="a9cd39277fb15d977d468435eb9b894f79f468233f0131509aa540ffda4f5953")
+
+    variant(
+        "build_type",
+        default="Release",
+        description="CMake build type",
+        values=("Debug", "Release"),
+    )
+
+    variant(
+        "model",
+        values=[
+            "std-data",
+            "std-indices",
+            "std-ranges",
+            "ocl",
+            "omp",
+            "hip",
+            "cuda",
+            "kokkos",
+            "sycl",
+            "sycl2020",
+            "acc",
+            "raja",
+            "tbb",
+            "thrust",
+        ],
+        default="std-data",
+        description="Specify the model to be built",
+    )
+
+    variant("flags", values=str, default=" ", description="Additional CXX flags to be provided")
+
+    variant(
+        "amd_arch", values=str, default="none", description="Target AMD GPU device being used."
+    )
+
+    variant(
+        "cuda_arch", values=str, default="none", description="The CUDA architecture being used."
+    )
```

Comment:
Can't we use `CudaPackage` and `RocmPackage` instead of duplicating the variants?

Reply:
Aren't these classes the whole package would need to inherit from? How do you select this based on the user choice of variant?

Reply:
Not sure I get what you're asking. Those two base classes are mixins, defining the `+cuda` variant etc. and a list of valid architecture for CUDA and ROCm (instead of accepting any string).

Reply:
Your Python is better than mine! Are you are suggesting we change `class Babelstream(CMakePackage)` to `class Babelstream(CMakePackage,CudaPackage,RocmPackage)`?

Does that not break the abstraction of the `model` variant here? I.e., we'd need to say `model=std-data` for that variant but spell it `+cuda` for the CUDA variant?

Further, how do we ensure that one-and-only-one variant is selected if we go via the mixins?

Reply:
@robj0nes can you follow up in any way? Thanks!

Reply:
I’m taking over progressing this. Is there an update to my question in the mixins?

Reply:
> Further, how do we ensure that one-and-only-one variant is selected if we go via the mixins?

You can use `conflicts` directives to disallow part of the search space. 

> Does that not break the abstraction of the model variant here? I.e., we'd need to say model=std-data for that variant but spell it +cuda for the CUDA variant?

I can't say much on the `model` variant, since this is package specific. It seems to me though that you can use conditional values and say the `cuda` value to `model` can be used only when `+cuda` is on:
```
        values=[
            "std-data",
            "std-indices",
            "std-ranges",
            "ocl",
            "omp",
            "hip",
            conditional("cuda", when="+cuda"),
            "kokkos",
            "sycl",
            "sycl2020",
            "acc",
            "raja",
            "tbb",
            "thrust",
        ],
```

Reply:
Thanks for the feedback.

The `model` option of the package as it is aims to provide identical functionality to the CUDA and ROCM mixins, but for *all* models. Spelling it `+cuda` but not `model=cuda` for CUDA, but `model=foo` for everything else is not sensible.

We're going to need to redesign the programming model selection to align with the mixin approach where that's supported, but provide the `+foo` notation where there is no mixin. I'm going to use the [Kokkos package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/kokkos/package.py) as inspiration for implementing this as there is similarity in the mechanics of choosing a back end programming model between the two packages.

Watch this space for a rewrite!

Reply:
What is the status of this PR?

Reply:
We are working on an update which uses the mixins. We have something working on some systems and are testing on others. We will push the update soon for your review.